packages/server/src/migrations/0.1.21-client-tokens.ts

```diff
@@ -0,0 +1,58 @@
+import { Migration } from '@botpress/messaging-engine'
+import { v4 as uuidv4 } from 'uuid'
+
+export class ClientTokensMigration extends Migration {
+  meta = {
+    name: ClientTokensMigration.name,
+    description: 'Transfers client tokens from msg_clients to a new table msg_client_tokens',
+    version: '0.1.21'
+  }
+
+  async valid() {
+    return this.trx.schema.hasTable('msg_clients')
+  }
+
+  async applied() {
+    return !(await this.trx.schema.hasColumn('msg_clients', 'token'))
+  }
+
+  async up() {
+    await this.trx.schema.createTable('msg_client_tokens', (table) => {
+      table.uuid('id').primary()
+      table.uuid('clientId').references('id').inTable('msg_clients')
+      table.string('token').notNullable()
+      table.timestamp('expiry').nullable()
+    })
+
+    const clients = await this.trx('msg_clients')
+    for (const client of clients) {
+      await this.trx('msg_client_tokens').insert({ id: uuidv4(), clientId: client.id, token: client.token })
+    }
+
+    await this.trx.schema.alterTable('msg_clients', (table) => {
+      table.dropColumn('token')
+    })
+  }
+
+  async down() {
+    await this.trx.schema.alterTable('msg_clients', (table) => {
+      table.string('token').unique()
+    })
+
+    const clients = await this.trx('msg_clients')
+
+    for (const client of clients) {
+      const clientTokens = await this.trx('msg_client_tokens').where({ clientId: client.id })
```

Comment:
Can clientTokens be empty?

Reply:
No why? It doesn't really make sense, you would just have a client that is inaccessible because it doesn't have any tokens. There's always going to be a master token that has all rigths.